core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedQueryScheduler.java

```diff
@@ -873,7 +874,8 @@ public static DistributedStagesScheduler create(
                 });
             }
 
-            Map<StageId, StageExecution> stageExecutions = new HashMap<>();
+            // Preserve topological ordering in stageExecutionsMap
```

Comment:
Because canceling a task will call `outputBuffer.destroy()`, which means that tasks consuming from said buffer can see `BufferResult(completed: true)` on the immediate subsequent request. This means that a task receiving a cancel before tasks consuming from it have themselves been canceled, aborted, or failed can trigger "finished" to propagate forward through the stages along worker to worker communication links unless the cancellation proceeds in topological order.

Essentially, when `DistributedStagesScheduler#cancel()` is called, we want to traverse the `stageExecutions` in topological order to reduce the possibility of that happening. That said, this issue shouldn't arise _now_ because `DistributedStagesScheduler#cancel()` is only called once the query state machine is finished anyway, but during development of these changes I did continue to see incorrect query results that seemed to stem from cancellations turning into incorrect query results, potentially due to tasks being sent both a cancel and abort command concurrently and those requests racing with one another to reach the to tasks (now fixed by `HttpRemoteTask` only allowing a single termination command to proceed).

Honestly, I'm still scratching my head a little bit about why tasks should call `outputBuffer.destroy()` on cancel instead of `outputBuffer.abort()`. It seems like that's a fragile behavior that can easily cause incorrect query results due to subtle concurrency or state machine bugs.

Reply:
Sorry, I mean "explain why this matters in the comment"... so future readers know why sorting is there.  You don't need to explain in this level of detail (but you could).  Generally, I like to see enough to get the reader thinking of the implications.